src/libraries/Microsoft.PowerFx.Core/Texl/Builtins/Collect.cs

```diff
@@ -0,0 +1,472 @@
+                //{
+                //    argType.DisplayNameProvider = datasourceType.DisplayNameProvider;
+                //}
+
+                // The subsequent args should all be aggregates.
+                if (!argType.IsAggregate)
+                {
+                    errors.EnsureError(args[i], TexlStrings.ErrBadType_Type, argType.GetKindString());
+                    fValid = false;
+                    continue;
+                }
+
+                // Promote the arg type to a table to facilitate unioning.
+                if (!argType.IsTable)
+                {
+                    argType = argType.ToTable();
+                }
+
+                // Checks if all record names exist against table type and if its possible to coerce.
+                bool checkAggregateNames = argType.CheckAggregateNames(datasourceType, args[i], errors, features, SupportsParamCoercion);
+                fValid = fValid && checkAggregateNames;
+
+                if (!itemType.IsValid)
+                {
+                    itemType = argType;
+                }
+                else
+                {
+                    var fUnionError = false;
+                    itemType = DType.Union(ref fUnionError, itemType, argType, useLegacyDateTimeAccepts: true, features);
+                    if (fUnionError)
+                    {
+                        errors.EnsureError(DocumentErrorSeverity.Severe, args[i], TexlStrings.ErrIncompatibleTypes);
+                        fValid = false;
+                    }
+                }
+
+                // We only support accessing entities in collections if the collection has only 1 argument that contributes to it's type
+                if (argc != 2 && itemType.ContainsDataEntityType(DPath.Root))
+                {
+                    fValid &= DropAllOfKindNested(ref itemType, errors, args[i], DKind.DataEntity);
+                }
+            }
+
+            Contracts.Assert(!itemType.IsValid || itemType.IsTable);
+            collectedType = itemType.IsValid ? itemType : DType.EmptyTable;
+            return fValid;
+        }
+
+        // Typecheck an invocation of Collect.
+        public override bool CheckTypes(CheckTypesContext context, TexlNode[] args, DType[] argTypes, IErrorContainer errors, out DType returnType, out Dictionary<TexlNode, DType> nodeToCoercedTypeMap)
+        {
+            Contracts.AssertValue(args);
+            Contracts.AssertAllValues(args);
+            Contracts.AssertValue(argTypes);
+            Contracts.Assert(args.Length == argTypes.Length);
+            Contracts.AssertValue(errors);
+            Contracts.Assert(MinArity <= args.Length && args.Length <= MaxArity);
+
+            var fValid = base.CheckTypes(context, args, argTypes, errors, out returnType, out nodeToCoercedTypeMap);
+
+            // Need a collection for the 1st arg
+            DType collectionType = argTypes[0];
+            if (!collectionType.IsTable)
```

Comment:
Triple check this ... this is where we could break PA.

Reply:
Collect always expects a table as arg0.